ribs/emitters/rankers.py

```diff
@@ -0,0 +1,368 @@
+"""Package with rankers for use across emitters.
+
+The rankers implemented in this file is intended to be used with emitters.
+Specifically, a ranker object should be initialized or passed in the emitters.
+The ``Ranker`` object will define the :meth:`rank` method which returns the
+result of an descending argsort of the solutions. It will also define a
+:meth:`reset` method which resets the internal states of the object.
+
+.. autosummary::
+    :toctree:
+
+    ribs.emitters.rankers.RandomDirectionRanker
+    ribs.emitters.rankers.TwoStageRandomDirectionRanker
+    ribs.emitters.rankers.ObjectiveRanker
+    ribs.emitters.rankers.TwoStageObjectiveRanker
+    ribs.emitters.rankers.ImprovementRanker
+    ribs.emitters.rankers.TwoStageImprovementRanker
+    ribs.emitters.rankers.RankerBase
+"""
+
+from abc import ABC, abstractmethod
+import numpy as np
+
+from ribs._docstrings import DocstringComponents, _core_docs
+
+__all__ = [
+    "RandomDirectionRanker",
+    "TwoStageRandomDirectionRanker",
+    "ObjectiveRanker",
+    "TwoStageObjectiveRanker",
+    "ImprovementRanker",
+    "TwoStageImprovementRanker",
+    "RankerBase",
+]
+
+# Define common docstrings
+_args = _core_docs["args"]
+_returns = DocstringComponents(
+    dict(index_batch="""
+    A batch of indicies representing a ranking of the solutions"""))
+
+
+class RankerBase(ABC):
+    """Base class for rankers.
+
+    Every ranker has a :meth:`rank` method that returns a list of indices
+    that indicate how the solutions should be ranked and a :meth:`reset` method
+    that resets the internal state of the ranker
+    (e.g. in :class:`ribs.emitters.rankers._random_direction_ranker`).
+
+    Child classes are only required to override :meth:`rank`.
+    """
+
+    @abstractmethod
+    def rank(self, emitter, archive, solution_batch, objective_batch,
+             measure_batch, metadata, add_statuses, add_values):
+        # pylint: disable=missing-function-docstring
+        pass
+
+    # Generates the docstring for rank
+    rank.__doc__ = f"""
+Generates a batch of indicies that represents an ordering of ``solution_batch``.
+
+Args:
+{_args.emitter}
```

Comment:
I don't think the definitions of `emitter` and `archive` from `_docstrings.py` fit here.